src/core/execute.c

```diff
@@ -2059,6 +2059,33 @@ static int apply_mount_namespace(Unit *u, const ExecContext *context,
         return r;
 }
 
+static int apply_working_directory(const ExecContext *context,
+                                   const ExecParameters *params,
+                                   const char *working_directory,
+                                   const bool needs_mount_ns) {
+
+        if (params->flags & EXEC_APPLY_CHROOT) {
+                if (!needs_mount_ns && context->root_directory)
+                        if (chroot(context->root_directory) < 0)
+                                return -errno;
+
+                if (chdir(working_directory) < 0 &&
+                    !context->working_directory_missing_ok)
+                        return -errno;
+
+        } else {
+                const char *d;
+
+                d = strjoina(strempty(context->root_directory), "/",
+                             strempty(working_directory));
+                if (chdir(d) < 0 &&
```

Comment:
It would be nicy to move the `if (chdir…)` part outside of the conditional. It's the same in both.